### DISCUSSION TITLE
Support for new device

### DISCUSSION THREAD

**Discussion by robio339** [syssi/esphome-config-examples#4] [General]
Hi Sebastian. I was planning to buy Victron smartshunt to get some idea for state of charge for my lifepo4 batteries, since JK bms state of charge tells us nothing useful. Than i found smart battery meter on Aliexpress. There were some reviews on YouTube compering it with smartshunt and results were looking promising. So i decided to buy one, since it was only 52eur including display.
https://www.aliexpress.com/item/1005005293243736.html?spm=a2g0o.order_list.order_list_main.230.21ef1802jcQ52S 
I went with KH140F which is 400A version.
It looks really nice with big display or in application on the phone and also state of charge look much more realistic than the ones from JK bms. 
I noticed in user manual there is chapter on communication protocol (page 26/pdf page 30)
If there is some time left in your day it would be really nice to get this gadget to send data to home assistant via Bluetooth. There is a thread for older KG model using RS-485 adapter but i didn't test it since i cant find RS-485 adapter so i ordered a new one.
Let me know if you could do you magic and if there is anything i can help - just not coding ;) .
Thanks
Robert

[Manual_en_zx.pdf](https://github.com/syssi/esphome-config-examples/files/11214645/Manual_en_zx.pdf)

**Comment by syssi** [comment]
Let's give it a try! Could you provide some photos of the device and the connector of the data cable?

**Reply by robio339** [reply]
Hi, this was really  fast. I can make some photos but there are great photos in manual. On page 10/pdf 14 there is a photo of communication connectors. They use RJ11 4pin connector. Connector no.5 goes to display and no.4 is for communication connection. Explanation for no.4 connector is:
(4).RS-485 communication connection interface
Used for protocol communication, and can also be used to connect another
measurement module for multi-machine communication. From left to right they are: B, A, GND, NC. 

Another question. Why cant i find "buy ma a coffee" button anymore on your github ?
Robert

**Reply by syssi** [reply]
Ping me as soon as your RS485 module did arrive. Please make sure to order the one with automatic flow control. You should find the "buy me a coffee" button at the top (last icon) of every README/project.

**Reply by syssi** [reply]
We will start using the example configuration:

https://github.com/syssi/esphome-config-examples/pull/5/files

**Comment by robio339** [comment]
I just find this coffee button. It really should be bigger ;) Logo on your first page is just the right size, the motive is right, just change text ;)

I will ones again try to find rs485 modules. I ordered few spare when i was setting up Epever mppt so they must be somewhere.
Thanks for now
Robert

**Reply by robio339** [reply]
Hi. i found one of the rs485 modules and tried the code. At first i got connected message on display every few seconds. I changed interval in code to 10 seconds and message come on every 10 seconds so on each interval. Then i remembered that there was some address settings in Junctek settings. I changed address from one to two and message went away. Since then i only got few overvoltage warnings on display at random times. We will see about this later if it will repeat.  I "decoded" logs and i can confirm that it reports correct values.
There are two logs attached. The one with number  1 at the end is from when address was set to 1 and the other with 2 at the end address was set to 2. The difference that i noticed is that there is no report for R00 when address is set to 2. Should i change to R00=2,2,1,\r\n? Anyway there are only device information in R00 which are not important to have in HA.
Let me know if there is something else i could do.
Thanks
Robi

[logs_junctek-kh-f_logs 1.txt](https://github.com/syssi/esphome-config-examples/files/11236782/logs_junctek-kh-f_logs.1.txt)
[logs_junctek-kh-f_logs 2.txt](https://github.com/syssi/esphome-config-examples/files/11236796/logs_junctek-kh-f_logs.2.txt)
![IMG_20230414_195506_886_resized(1)](https://user-images.githubusercontent.com/30301471/232150195-e3b0716a-71e7-4b1d-8aab-1a5e9cd5760a.jpg)
![IMG_20230414_202205_592_resized(1)](https://user-images.githubusercontent.com/30301471/232150211-7dbe2462-01ac-4c02-bfe1-d22c39e8f933.jpg)

**Reply by syssi** [reply]
I'm bit confused about the order of the log messages. I've updated the test YAML slightly:

https://github.com/syssi/esphome-config-examples/pull/5/commits/0ad5b0741a4c4a59bece23d645fa5546f546502a

Could you provide another log for address 1?

**Reply by robio339** [reply]
Should there be any other change in YAML other then 10s?
Here is new log with address set to 1
Thanks
Robi
[logs_junctek-kh-f_logs (2).txt](https://github.com/syssi/esphome-config-examples/files/11237206/logs_junctek-kh-f_logs.2.txt)

**Reply by syssi** [reply]
Okay, the order of the messages is fine but the amount is confusing. Are there more RS485 devices on the bus? Did you run the Android app simultaneously? If you stop the Android app do you see a log message per second / per request instead of around 5 foreign requests and responses per second?

**Reply by syssi** [reply]
I try to rephrase my observation:

`<<<` indicates incoming traffic
`<<<` indicates outgoing traffic

The log shows a lot of these lines:

```
<<< ":R50=1,\n"
<<< ":r50=1,121,5299,0,92035,0,187577,381383,115,0,99,0,4779,100,230414,232049,\r\n"
<<< ":R51=1,\n"
<<< ":r51=1,82,0,0,0,0,0,160,0,0,1000,100,100,100,0,0,1,100,0,0,0,20,20,255,0,0,0,10,\r\n"
<<< ":R50=1,\n"
<<< ":r50=1,123,5299,0,92035,0,187577,381384,115,0,99,0,4779,100,230414,232050,\r\n"
```

The request (`:R50=1,\n`) and the response (`:r50=1,121,5299,...`) are incoming data. I've expected silence on the bus expect if we send a request:

```
[23:20:59][D][uart_debug:158]: >>> ":R00=1,2,1,\r\n"
[23:20:59][D][uart_debug:158]: <<< ":r00=1,197,4140,121,15,\r\n"

[23:21:00][D][uart_debug:158]: >>> ":R50=1,2,1,\r\n"
[23:21:00][D][uart_debug:158]: <<< ":r50=1,135,5299,0,92035,0,187577,381390,115,0,99,0,4779,100,230414,232056,\r\n"

[23:21:01][D][uart_debug:158]: >>> ":R51=1,2,1,\r\n"
[23:21:01][D][uart_debug:158]: <<< ":r51=1,82,0,0,0,0,0,160,0,0,1000,100,100,100,0,0,1,100,0,0,0,20,20,255,0,0,0,10,\r\n"
```

Our request payload (`:R50=1,2,1,\r\n`) is a bit different to the request payload (`:R50=1,\n`) of the foreign bus consumer.

**Reply by robio339** [reply]
Hi, good morning. Display is also communicating over rs485. Here is new log. I disconnected display cable. I opened app on phone and set address to 2 and than i run log. Now you can see at start esp sending request but no replay. Than i change address in app to 1 and there comes replay. App was on all the time and as you can see interval is 10s. 
I guess that is why address needs to be changed. There is this written in instructions:

15).Address matching
In the system settings interface, press the 【▲】【▼】keys to switch to the address matching setting. Short press the 【OK】key to select the value with the cursor. Use the 【▲】【▼】keys to switch the communication address, and press the 【OK】key again to confirm the communication address between the display module and the measurement module. The communication address is set to P01 by default. When multiple measurement modules are connected, switching the current
communication address can control the corresponding measurement module
This setting i was changing from 1 to 2 and back.
and

25).Display host mode (Display screen communication)
In the system settings interface, press the 【▲】【▼】keys to switch to the display host mode. Short press the 【OK】key to select open or close with the cursor. Use the 【▲】【▼】keys to switch, and press the 【OK】key again to confirm whether to turn on or off the display screen communication. It is necessary to turn on the display screen communication function when using communication protocol communication. (After turning off the display screen communication function, the display module will not update data automatically.)
This setting is set to on.

[logs_junctek-kh-f_logs (3).txt](https://github.com/syssi/esphome-config-examples/files/11238828/logs_junctek-kh-f_logs.3.txt)

Robi

**Reply by syssi** [reply]
Alright! In other words: Polling the device is optional if the display is attached and configured as host/master. The ESPHome component should listen passively if the display is the bus master.

**Reply by robio339** [reply]
Hi. just did another log. I removed from code whole interval section. I plugged back display and esp receives information even there is no request sent. If i disconnect display logging stops if i plug it back it starts again.
[logs_junctek-kh-f_logs (3).txt](https://github.com/syssi/esphome-config-examples/files/11239190/logs_junctek-kh-f_logs.3.txt)

**Reply by syssi** [reply]
Please try the external component of tfyoung as next step. I've updated the YAML already:

https://github.com/syssi/esphome-config-examples/blob/kh140f/esp32-junctek-kh-f.yaml

**Reply by robio339** [reply]
It works. Display is connected and for now no  error messages are shown .
[logs_junctek-kh-f-dummy_logs.txt](https://github.com/syssi/esphome-config-examples/files/11239267/logs_junctek-kh-f-dummy_logs.txt)
![image](https://user-images.githubusercontent.com/30301471/232214815-16dd68d5-54e1-4720-affd-36f7b3338c55.png)

**Reply by syssi** [reply]
Cool! I've prepared some changes of the external component and hope the author will respond. Do you miss something important?

**Reply by robio339** [reply]
Hi. I was next to display for 20 minutes and there are still error messages coming at random times. Sometimes its empty then low voltage protection, then low temp, then high temp. Just random messages  at random times. I opened log and watched if there is any repeated event in log when error accrues but i didn't see a pattern. Sometimes no 5seconds sometimes after few minutes. There are some red error messages in log (check on the end of the log) but it doesn't look to be connected with error on display. 

[logs_junctek-kh-f-dummy_logs (1).txt](https://github.com/syssi/esphome-config-examples/files/11239542/logs_junctek-kh-f-dummy_logs.1.txt)

So we have now battery current in out, voltage, temperature, state of charge. Useful things would also be:
-remain ah
-power
-charging energy 
-discharging energy
- if possible relay switch on off

Settings can easily be done over app. The only useful setting for me to have it in HA would maybe be some things in connection with controlling relay . There is output for relay from Junctek . It can be set to be triggered with protection parameters. The only problem is that if i understood right, when triggered it can only be canceled by time and not value. And also if this random errors will stay maybe relay function wont be useful. So lets wait with this.
I will find one relay and try what is happening with it during errors.
Thanks 
Robert
 
![315202315416](https://user-images.githubusercontent.com/30301471/232227663-4e76d364-4132-46df-9567-2e7ddd65f68a.jpg)

**Reply by syssi** [reply]
I've removed the polling from the external component to avoid confusing the display. Please update the YAML from:

```
external_components_source: github://tfyoung/esphome-junctek_kgf@main
```

to

```
external_components_source: github://syssi/esphome-junctek_kgf@dont-poll
```

and give it another try!

**Reply by robio339** [reply]
Entities are gone. They become unavailable
[logs_junctek-kh-f-dummy_logs (2).txt](https://github.com/syssi/esphome-config-examples/files/11239889/logs_junctek-kh-f-dummy_logs.2.txt)

**Reply by syssi** [reply]
Does it work again if you restart the node a few times?

**Reply by syssi** [reply]
I was able to reproduce the issue. I have an idea whats going on.

**Reply by robio339** [reply]
Ok, restarting didn't make no difference.

**Reply by syssi** [reply]
I have to debug the receiver implementation. This will take some time. I will get back to you!

**Reply by syssi** [reply]
I've pushed a fix the the `dont-poll` branch. Please update your ESP and give it another try!

**Reply by robio339** [reply]
Hi, yes entities are back. I went and sat down by display for around 10 minutes and no errors on display. So good work 
Here is the latest log

[logs_junctek-kh-f-dummy_logs (3).txt](https://github.com/syssi/esphome-config-examples/files/11241936/logs_junctek-kh-f-dummy_logs.3.txt)
